### ISSUE TITLE
Bug: Annotation @javax.validation.constraints.NotNull not correctly parsed

### ISSUE THREAD

**Issue by ghilainm** [INRIA/spoon#1773]
Maven dependencies:
```
<dependency>
    <groupId>javax.validation</groupId>
    <artifactId>validation-api</artifactId>
   <version>2.0.0.Final</version>
</dependency>
 <dependency>
   <groupId>fr.inria.gforge.spoon</groupId>
    <artifactId>spoon-core</artifactId>
    <version>6.0.0</version>
</dependency>
```

It seems like the field parsing is not working with this specific annotation (others seems to be working fine!)
Here is the field: 
@NotNull
public String mandatoryField;

Here is the toString() of the spoon model constructed:
```
public java.lang.@javax.validation.constraints.NotNull
String mandatoryField;
```

It should be this according t me:
```
@javax.validation.constraints.NotNull
public java.lang.String mandatoryField;
```

I don't understand how this is possible. This is working fine for example with annotation @javax.xml.bind.annotation.XmlTransient

**Comment by surli** [comment]
Hi @ghilainm 

Thanks for reporting it, I'll have a look.

**Comment by surli** [comment]
OK after investigating it seems that this bug is not a bug. 

Here the NotNull annotation can be applied on a type or on a field (TYPE_USE and FIELD in targets). 
And in JLS it is written that:

> `@Foo int f;`
> @Foo is a declaration annotation on f if Foo is meta-annotated by @Target(ElementType.FIELD), and **a type annotation on int if Foo is meta-annotated by @Target(ElementType.TYPE_USE)**. It is possible for @Foo to be both a declaration annotation and a type annotation simultaneously. 

(Source: https://docs.oracle.com/javase/specs/jls/se9/html/jls-9.html#jls-9.7.3)

So here we attached `@NotNull` to the type String. 
Now when we pretty print it with fully qualified name, we cannot pretty print 
`@javax.validation.constraints.NotNull java.lang.String`

according to the JLS, because:
> For example, assume an annotation type TA which is meta-annotated with just @Target(ElementType.TYPE_USE). The terms @TA java.lang.Object and java.@TA lang.Object are illegal because the simple name to which @TA is closest is classified as a package name. On the other hand, java.lang.@TA Object is legal. 

(Source: https://docs.oracle.com/javase/specs/jls/se9/html/jls-9.html#jls-9.7.4 - end of the section)

So the only solution here is to pretty print 
```
public java.lang.@javax.validation.constraints.NotNull String mandatoryField
```

But I agree it's a tricky one.

**Comment by ghilainm** [comment]
I must say I don't understand everything :). Probably my use case is wrong then. If it can be attached to both elements I should find it also in the annotations no ?

**Comment by tdurieux** [comment]
> So here we attached @NotNull to the type String.
Now when we pretty print it with fully qualified name, we cannot pretty print
@javax.validation.constraints.NotNull java.lang.String

I don't agree, in this case, it should be:
`@javax.validation.constraints.NotNull String` because we should not print the fully qualified name of `java.lang.*` element

> I must say I don't understand everything :). Probably my use case is wrong then. If it can be attached to both elements I should find it also in the annotations no ?

Apparently, the annotation `NotNull` is attached to the return type `String` of the method and not on the method.
In order to keep this link between the annotation and the return type in fully qualified name method, we have to print the annotation just before the name of the class (here String).

**Comment by ghilainm** [comment]
Okay, how can I attach it to the method and not to the return type? Seems like a silly question but it is not obvious for me :). I would like this annotation to be available of the method like the other method annotations.

**Comment by surli** [comment]
> Okay, how can I attach it to the method and not to the return type?

In fact after reading back the JLS it's specified that 
> It is possible for @foo to be both a declaration annotation and a type annotation simultaneously.

It's not absolutely clear but it seems to suggest that in that case the annotation should be attached to the type AND to the field (or the method). With this interpretation we indeed have a bug in Spoon. WDYT @tdurieux ?

> I don't agree, in this case, it should be:
@javax.validation.constraints.NotNull String because we should not print the fully qualified name of java.lang.* element

This is another debate about printing or not java.lang in Spoon: here I used the spec defined by @GerardPaligot in #520.

**Comment by ghilainm** [comment]
Do you know in which release this will be included?

**Comment by surli** [comment]
> Do you know in which release this will be included and when this released is planned?

It will be included in the next release, it should be Spoon 6.1.0. Usually we release around every 3 monthes, so next release should be done in february or march. 

In the meantime you can use the snapshot of Spoon 6.1.0 if you like.

**Comment by surli** [comment]
@ghilainm FYI we finally released Spoon 6.1.0 this week, so you should be able to use it with the bug fix :)

**Comment by ghilainm** [comment]
@surli 
Thx a lot man ;). I'll check it when I have some time :p. I implemented work around in the mean time.
